src/Build/Graph/ProjectGraph.cs

```diff
@@ -54,5 +59,139 @@ public ProjectGraph(string entryProjectFile)
+        /// <returns>A dictionary containing the target list for each node.</returns>
+        public IReadOnlyDictionary<ProjectGraphNode, string[]> GetTargetLists(string[] entryProjectTargets)
+        {
+            // Seed the dictionary with empty lists for every node.
+            Dictionary<ProjectGraphNode, List<string>> targetLists = _projectNodes.ToDictionary(node => node, node => new List<string>());
+
+            // Initial state of the graph traversal.
+            var entryEdge = new ProjectGraphNodeWithTargets(EntryProjectNode, entryProjectTargets);
+            var visitedEdges = new HashSet<ProjectGraphNodeWithTargets>();
+            visitedEdges.Add(entryEdge);
+            var edgesToVisit = new Queue<ProjectGraphNodeWithTargets>();
+            edgesToVisit.Enqueue(entryEdge);
+
+            // Traverse the entire graph, visiting each node once.
+            while (edgesToVisit.Count > 0)
+            {
+                ProjectGraphNodeWithTargets nodeWithTargets = edgesToVisit.Dequeue();
+                ProjectGraphNode node = nodeWithTargets.Node;
+                IList<string> targets = nodeWithTargets.Targets;
+
+                targetLists[node].AddRange(targets);
+
+                // No need to continue if this node has no project references.
+                if (node.ProjectReferences.Count == 0)
+                {
+                    continue;
+                }
+
+                // Based on the entry points of this project, determine which targets to propagate down to project references.
+                var targetsToAppendToChildren = new List<string>();
+                ICollection<ProjectItem> projectReferenceTargets = node.Project.GetItems(ProjectReferenceTargetsItemType);
+                foreach (var target in targets)
+                {
+                    // TODO: Normalize targets? Default targets needs expansion. Flip the loop maybe
+
+                    foreach (var projectReferenceTarget in projectReferenceTargets)
+                    {
+                        if (projectReferenceTarget.EvaluatedInclude.Equals(target, StringComparison.OrdinalIgnoreCase))
+                        {
+                            string targetsMetadataValue = projectReferenceTarget.GetMetadataValue(ProjectReferenceTargetsMetadataName);
+                            targetsToAppendToChildren.AddRange(ExpressionShredder.SplitSemiColonSeparatedList(targetsMetadataValue));
+                        }
+                    }
+                }
+
+                // Queue the project references for visitation, if the edge hasn't already been traversed.
+                foreach (var projectReference in node.ProjectReferences)
+                {
+                    var projectReferenceEdge = new ProjectGraphNodeWithTargets(projectReference, targetsToAppendToChildren);
+                    if (visitedEdges.Add(projectReferenceEdge))
```

Comment:
Else branch means cycle right? If cycle, this should error.

Reply:
If cycles would be valid, we'd need to switch to a `while(new changes){PropagateEdgeInfo}` visitation to account for feedback.

---
In reply to: [218885297](https://github.com/Microsoft/msbuild/pull/3769#discussion_r218885297) [](ancestors = 218885297)

Reply:
Cycles are not valid. We will want to detect cycles on graph creation though (see #3757), so at this point it should be safe to assume no cycles.

The `else` here isn't necessarily a cycle though. Edges are defined by the incoming target list and the project, but *not* the referencing project. So if two projects share a project reference, those edges will be deduped by this. So these aren't distinct edges really, just incoming edges from the POV of a given project. It's an optimization so that projects who many many other projects depend on don't need to have a massive target list prior to target list deduplication below.